src/traces/scattermapbox/plot.js

```diff
@@ -56,72 +74,108 @@ proto.setSourceData = function(k, opts) {
+                k = order[i];
+                map.removeLayer(this.layerIds[k]);
+            }
+            for(i = 0; i < order.length; i++) {
+                k = order[i];
+                opts = optsAll[k];
+                this.addLayer(k, opts, below);
+            }
+            this.below = below;
+        }
+    } else if(trace.cluster.enabled && !this.clusterEnabled) {
+        for(i = ORDER.nonCluster.length - 1; i >= 0; i--) {
+            k = ORDER.nonCluster[i];
             map.removeLayer(this.layerIds[k]);
+            map.removeSource(this.sourceIds[k]);
         }
-        for(i = 0; i < ORDER.length; i++) {
-            k = ORDER[i];
+        this.addSource('circle', optsAll.circle, trace.cluster);
+        for(i = 0; i < ORDER.cluster.length; i++) {
+            k = ORDER.cluster[i];
             opts = optsAll[k];
             this.addLayer(k, opts, below);
         }
-        this.below = below;
-    }
-
-    for(i = 0; i < ORDER.length; i++) {
-        k = ORDER[i];
-        opts = optsAll[k];
-
-        subplot.setOptions(this.layerIds[k], 'setLayoutProperty', opts.layout);
-
-        if(opts.layout.visibility === 'visible') {
-            this.setSourceData(k, opts);
-            subplot.setOptions(this.layerIds[k], 'setPaintProperty', opts.paint);
+        this.clusterEnabled = trace.cluster.enabled;
+    } else if(!trace.cluster.enabled && this.clusterEnabled) {
+        for(i = 0; i < ORDER.cluster.length; i++) {
+            k = ORDER.cluster[i];
+            map.removeLayer(this.layerIds[k]);
+        }
+        map.removeSource(this.sourceIds.circle);
+        for(i = 0; i < ORDER.nonCluster.length; i++) {
+            k = ORDER.nonCluster[i];
+            opts = optsAll[k];
+            this.addSource(k, opts, trace.cluster);
+            this.addLayer(k, opts, below);
         }
+        this.clusterEnabled = trace.cluster.enabled;
     }
 
-    // link ref for quick update during selections
+  // link ref for quick update during selections
     calcTrace[0].trace._glTrace = this;
 };
 
 proto.dispose = function dispose() {
     var map = this.subplot.map;
-
-    for(var i = ORDER.length - 1; i >= 0; i--) {
-        var k = ORDER[i];
+    var order = this.clusterEnabled ? ORDER.cluster : ORDER.nonCluster;
+    for(var i = order.length - 1; i >= 0; i--) {
+        var k = order[i];
         map.removeLayer(this.layerIds[k]);
         map.removeSource(this.sourceIds[k]);
     }
 };
 
 module.exports = function createScatterMapbox(subplot, calcTrace) {
     var trace = calcTrace[0].trace;
-    var scatterMapbox = new ScatterMapbox(subplot, trace.uid);
+    var scatterMapbox = new ScatterMapbox(
+    subplot,
+    trace.uid,
+    trace.cluster.enabled
+  );
     var optsAll = convert(subplot.gd, calcTrace);
-    var below = scatterMapbox.below = subplot.belowLookup['trace-' + trace.uid];
+    var below = (scatterMapbox.below = subplot.belowLookup['trace-' + trace.uid]);
```

Comment:
Please drop the parentheses here:
```suggestion
    var below = scatterMapbox.below = subplot.belowLookup['trace-' + trace.uid];
```